### PR TITLE
Network approach to BP algorithm

### PR DESCRIPTION
**Pull request by elshize**

### CODE REVIEW

File: include/recursive_graph_bisection.hpp
Comment (amallia) [inline include/recursive_graph_bisection.hpp]: Really needed?

File: include/recursive_graph_bisection.hpp
Comment (amallia) [inline include/recursive_graph_bisection.hpp]: An assert will produce the same result here.

File: include/recursive_graph_bisection.hpp
Comment (elshize) [inline include/recursive_graph_bisection.hpp]: This is somewhat temporary. I needed some assertion at hoc, didn't want to put it anywhere else, since I know you've looked into assertions and maybe already planned to do something else.

**Review by amallia — CHANGES_REQUESTED** [review]

File: include/recursive_graph_bisection.hpp
Comment (elshize) [reply include/recursive_graph_bisection.hpp]: See comment

File: include/recursive_graph_bisection.hpp
Comment (amallia) [reply include/recursive_graph_bisection.hpp]: Oh ok, read only now. Just commented about that. Thanks for clarifying

File: include/recursive_graph_bisection.hpp
Comment (elshize) [reply include/recursive_graph_bisection.hpp]: An assert will not print the values.

Comment (amallia) [reply include/recursive_graph_bisection.hpp]: Shall we try https://github.com/amallia/better_assert ?

File: include/recursive_graph_bisection.hpp
Comment (elshize) [reply include/recursive_graph_bisection.hpp]: Actually, let me remove all of this for now, we should add it separately globally somewhere. Just needed something to debug and started playing around.

File: src/recursive_graph_bisection.cpp
Comment (amallia) [inline src/recursive_graph_bisection.cpp:71]: Why does ordering need to be printed if a `.mapping` file is generated?

**Review by amallia — APPROVED** [review]
LGTM

File: src/recursive_graph_bisection.cpp
Comment (elshize) [reply src/recursive_graph_bisection.cpp:71]: Frankly, I didn't notice (or remember) that it it was printing out the mapping. Which makes me think that maybe that should be a separate function called from the source file. Just something to think about.

Anyway, I think it still makes sense to print it out as text, might be handy for analysis or debugging. But I can remove it if it bothers you.

Comment (amallia) [reply src/recursive_graph_bisection.cpp:71]: We can leave it. I don't mind. I was just wondering

File: include/recursive_graph_bisection.hpp
Comment (elshize) [reply include/recursive_graph_bisection.hpp]: Sure, why not.

**Comment by elshize** [comment]
How come `constexpr` fails now on Travis? Did it work before?

**Comment by amallia** [comment]
I got rid of C++17, but feel free to re-add it in CMakeLists.txt

**Comment by elshize** [comment]
Oh, so you never ran the BP on Travis before? I thought you had. Anyway, I'll deal with it.

**Comment by amallia** [comment]
Yes, I did. I just removed all the `if constexpr` as I don't need it anymore.

**Comment by elshize** [comment]
So where are these changes? Did you work on another branch? Coz nothing was on `master` and this branch had `constexpr`s before my changes.

**Comment by amallia** [comment]
I will update the compiler in travis. It is using GCC-6, so there is not support for `if constexpr`. My bad!

https://github.com/pisa-engine/pisa/pull/11

**Comment by amallia** [comment]
Updated to use GCC 7

**Comment by elshize** [comment]
There seems to be a problem with `QMX` library that a test is added but no binary and `ctest` fails with "Not Run". We should fix that before merging this.

**Comment by amallia** [comment]
Where is `qmxTest` coming from? I never created this test AFAIK. I will give it a look tonight.

EDIT: coming from here https://github.com/amallia/QMX/blob/1c5c7385de244589f04c30bfe9ec12de0c99c15e/test/qmxTest.cpp

**Comment by amallia** [comment]
Re-running CI in order to squash and merge

**Comment by amallia** [comment]
`C++17` is blocked by https://github.com/stxxl/stxxl/pull/94

**Comment by elshize** [comment]
> `C++17` is blocked by [stxxl/stxxl#94](https://github.com/stxxl/stxxl/pull/94)

What do you use from `stxxl`? I guess we can for now use c++14 globally and only set c++17 wherever we need.

Another question: how come it compiles? I set c++17 in another branch for the entire project, and I found no issues.

**Comment by amallia** [comment]
Good point @elshize! Added C++17 only for `recursive_graph_bisection` and set C++14 globally.

The problem is that Clang has implemented C++17 and it has removed `random_shuffle` and other stuff, according to the new standard. GCC 7 (used by Travis) does not have this problem, but it will show up soon.

**Comment by elshize** [comment]
Oh, it's a clang-specific issue, got it.

**Pull request merged**
